### PR TITLE
Toric code:  Add optimizers for toric_code (#6)

### PR DESCRIPTION
**Pull request by dstrain115**
Adds optimizers (transformers) to the toric code directory in ReCirq. 

These transformers prepare a toric code circuit for hardware execution.  Among the transformers in this file are:
- Tools to convert CNOT to CZ gates.
- Tools to add spin echos (pairs of X gates) to the circuit during longer idle periods to improve frequency.
- Tools to move around (defer) single qubits gates to reduce circuit depth.

Note: these transformers are still as free functions and will need to use the new transformers API in a later PR.

### CODE REVIEW

File: recirq/toric_code/optimizers.py [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+    if moment_on_deck:
+        new_circuit += moment_on_deck
+
+    return new_circuit
+
+
+def defer_single_qubit_gates(circuit: cirq.Circuit) -> cirq.Circuit:
+    """Create a new circuit where some early single-qubit gates have been pushed later.
+
+    This is useful if the qubits begin the circuit in |0⟩; we leave them in the initial state
```
Comment (mpharrigan) [inline recirq/toric_code/optimizers.py:190]: this sentence is grammatically confusing

File: recirq/toric_code/optimizers.py [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+    """
+    new_circuit = copy.deepcopy(circuit)
+
+    exclusively_1q_moments: List[
+        int
+    ] = []  # We can move 1q gates in/out of these moments
+    deferrable_qubits = set(new_circuit.all_qubits())
+    qubit_to_deferred_ops: Dict[cirq.Qid, List[cirq.Operation]] = defaultdict(
+        lambda: []
+    )
```
Comment (mpharrigan) [inline recirq/toric_code/optimizers.py:206]: `defaultdict(list)` is concise and idiomatic

**Review by mpharrigan — COMMENTED** [review]
looks good, after a quick review

File: recirq/toric_code/optimizers.py [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+    if cirq.qid_shape(gate0) != cirq.qid_shape(gate1):
+        return False
+    # Evaluate unitaries
+    return cirq.allclose_up_to_global_phase(
+        cirq.unitary(gate0), cirq.unitary(gate1), atol=UNITARY_ATOL
+    )
+
+
+def _qubits_with_hadamard(moment: cirq.Moment) -> Set[cirq.Qid]:
+    return {op.qubits[0] for op in moment if _gates_are_close(op.gate, cirq.H)}
```
[the remark this answers is not in the document]
Comment (dstrain115) [reply recirq/toric_code/optimizers.py:38]: Done.

File: recirq/toric_code/optimizers.py (hunk 2 above)
Comment (dstrain115) [reply recirq/toric_code/optimizers.py]: Done.

File: recirq/toric_code/optimizers.py (hunk 1 above)
Comment (dstrain115) [reply recirq/toric_code/optimizers.py]: Changed.

File: recirq/toric_code/optimizers.py [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+    Note that this excludes measurement."""
+    return all(len(op.qubits) == 1 and cirq.has_unitary(op) for op in moment)
+
+
+def _has_cnot(moment: cirq.Moment) -> bool:
+    """Returns True if any operations in the moment are CNOT."""
+    return any(_gates_are_close(op.gate, cirq.CNOT) for op in moment)
+
+
+# Helper functions for decomposing operations
```
[the remark this answers is not in the document]
Comment (dstrain115) [reply recirq/toric_code/optimizers.py]: I just deleted the line comments.  This file isn't that long, so I am not going to split it up.  I need to go back and fix this to use Transformers API anyway, so I will reconsider then.

**Comment by mpharrigan** [comment]
General comment: I think in the future we want to move away from the term "optimizer" for circuit transformation / compilation cc @tanujkhattar . It's a highly-overloaded term in our space.

Can you add a short PR description?

File: recirq/toric_code/optimizers.py [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+                    ) from ex
+
+            new_moment_ops.append(new_op)
+
+        modified_circuit.append(cirq.Moment(new_moment_ops))
+
+    return modified_circuit, pauli_frame / pauli_frame.coefficient
+
+
+def insert_echos_on_active_qubits(
```
Comment (mpharrigan) [inline recirq/toric_code/optimizers.py:306]: what's an active qubit? like one that's in the provided `circuit`? Presumably the active qubits are indeed idle during the to-be-inserted echo phase

**Review by mpharrigan — APPROVED** [review]

File: recirq/toric_code/optimizers.py (hunk 5 above)
Comment (dstrain115) [reply recirq/toric_code/optimizers.py]: I think active qubits means the ones used in the toric code, but I agree it is a confusing wording.  Changed to idle qubits through this file.

**Review by mpharrigan — APPROVED** [review]

**Pull request merged**